src/compiler/types.rs

```diff
@@ -1069,11 +1075,219 @@ impl From<FreshableIntrinsicType> for Type {
+        Type::LiteralType(literal_type)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BaseLiteralType {
+    _type: BaseType,
+    fresh_type: WeakSelf<Type>,
+    regular_type: WeakSelf<Type>,
+}
+
+impl BaseLiteralType {
+    pub fn new(type_: BaseType) -> Self {
+        Self {
+            _type: type_,
+            fresh_type: WeakSelf::new(),
+            regular_type: WeakSelf::new(),
+        }
+    }
+}
+
+impl TypeInterface for BaseLiteralType {
+    fn flags(&self) -> TypeFlags {
+        self._type.flags()
+    }
+}
+
+impl LiteralTypeInterface for BaseLiteralType {
+    fn fresh_type(&self) -> Option<&Weak<Type>> {
+        self.fresh_type.try_get()
+    }
+
+    fn set_fresh_type(&self, fresh_type: &Rc<Type>) {
+        self.fresh_type.init(fresh_type, false);
+    }
+
+    fn get_or_initialize_fresh_type(
+        &self,
+        type_checker: &TypeChecker,
+        wrapper: &Rc<Type>,
+    ) -> Rc<Type> {
+        panic!("Shouldn't call get_or_initialize_fresh_type() on base BaseLiteralType");
+    }
+
+    fn regular_type(&self) -> Rc<Type> {
+        self.regular_type.get().upgrade().unwrap()
+    }
+
+    fn set_regular_type(&self, regular_type: &Rc<Type>) {
+        self.regular_type.init(regular_type, false);
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct NumberLiteralType {
+    _literal_type: BaseLiteralType,
+    value: Number,
+}
+
+impl NumberLiteralType {
+    pub fn new(literal_type: BaseLiteralType, value: Number) -> Self {
+        Self {
+            _literal_type: literal_type,
+            value,
+        }
+    }
+
+    fn create_fresh_type_from_self(
+        &self,
+        type_checker: &TypeChecker,
+        wrapper: &Rc<Type>,
+    ) -> Rc<Type> {
+        let fresh_type = type_checker.create_number_literal_type(
+            self.flags(),
+            self.value,
+            Some(wrapper.clone()),
+        );
+        match &*fresh_type {
+            Type::LiteralType(literal_type) => {
+                literal_type.set_fresh_type(&fresh_type);
+            }
+            _ => panic!("Expected LiteralType"),
+        }
+        self.set_fresh_type(&fresh_type);
+        self.fresh_type().unwrap().upgrade().unwrap()
+    }
+}
+
+impl TypeInterface for NumberLiteralType {
+    fn flags(&self) -> TypeFlags {
+        self._literal_type.flags()
+    }
+}
+
+impl LiteralTypeInterface for NumberLiteralType {
+    fn fresh_type(&self) -> Option<&Weak<Type>> {
+        self._literal_type.fresh_type()
+    }
+
+    fn set_fresh_type(&self, fresh_type: &Rc<Type>) {
+        self._literal_type.set_fresh_type(fresh_type);
+    }
+
+    fn get_or_initialize_fresh_type(
+        &self,
+        type_checker: &TypeChecker,
+        wrapper: &Rc<Type>,
+    ) -> Rc<Type> {
+        if self.fresh_type().is_none() {
+            let fresh_type = self.create_fresh_type_from_self(type_checker, wrapper);
+            self.set_fresh_type(&fresh_type);
+            return self.fresh_type().unwrap().upgrade().unwrap();
```

Comment:
I believe that it manages to "squeak by" because we return a strong reference which then immediately gets cached into `number_literal_types`, so the strong reference count never drops to 0